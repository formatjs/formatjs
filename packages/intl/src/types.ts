import {
  IntlMessageFormat,
  Formats,
  FormatXMLElementFn,
  FormatError,
  PrimitiveType,
  Options as IntlMessageFormatOptions,
} from 'intl-messageformat'
import {DateTimeFormat} from '@formatjs/ecma402-abstract'
import {MessageFormatElement} from '@formatjs/icu-messageformat-parser'
import IntlListFormat, {
  IntlListFormatOptions,
  Part,
} from '@formatjs/intl-listformat'
import {DisplayNames, DisplayNamesOptions} from '@formatjs/intl-displaynames'
import {
  MissingTranslationError,
  MessageFormatError,
  MissingDataError,
  InvalidConfigError,
  UnsupportedFormatterError,
} from './error'
import {DEFAULT_INTL_CONFIG} from './utils'
import {NumberFormatOptions} from '@formatjs/ecma402-abstract'

declare global {
  namespace FormatjsIntl {
    interface Message {}
    interface IntlConfig {}
  }
}

type MessageIds = FormatjsIntl.Message extends {ids: string}
  ? FormatjsIntl.Message['ids']
  : string

type Locale = FormatjsIntl.IntlConfig extends {locale: string}
  ? FormatjsIntl.IntlConfig['locale']
  : string

export type OnErrorFn = (
  err:
    | MissingTranslationError
    | MessageFormatError
    | MissingDataError
    | InvalidConfigError
    | UnsupportedFormatterError
    | FormatError
) => void

export type OnWarnFn = (warning: string) => void

/**
 * Config for intl object.
 * Generic type T is the type of potential rich text element. For example:
 * With React, T would be React.ReactNode
 */
export interface ResolvedIntlConfig<T = string> {
  locale: Locale
  timeZone?: string
  fallbackOnEmptyString?: boolean
  formats: CustomFormats
  messages:
    | Record<MessageIds, string>
    | Record<MessageIds, MessageFormatElement[]>
  defaultLocale: string
  defaultFormats: CustomFormats
  defaultRichTextElements?: Record<string, FormatXMLElementFn<T>>
  onError: OnErrorFn
  onWarn?: OnWarnFn
}

export interface CustomFormats extends Partial<Formats> {
  relative?: Record<string, Intl.RelativeTimeFormatOptions>
}

export interface CustomFormatConfig {
  format?: string
}

export type FormatDateOptions = Omit<
  Intl.DateTimeFormatOptions,
  'localeMatcher'
> &
  CustomFormatConfig
export type FormatNumberOptions = Omit<NumberFormatOptions, 'localeMatcher'> &
  CustomFormatConfig
export type FormatRelativeTimeOptions = Omit<
  Intl.RelativeTimeFormatOptions,
  'localeMatcher'
> &
  CustomFormatConfig
export type FormatPluralOptions = Omit<
  Intl.PluralRulesOptions,
  'localeMatcher'
> &
  CustomFormatConfig

export type FormatListOptions = Omit<IntlListFormatOptions, 'localeMatcher'>

export type FormatDisplayNameOptions = Omit<
  DisplayNamesOptions,
  'localeMatcher'
>

export interface IntlFormatters<T = any, R = T> {
  formatDateTimeRange(
    from: Parameters<DateTimeFormat['formatRange']>[0],
    to: Parameters<DateTimeFormat['formatRange']>[1],
    opts?: FormatDateOptions
  ): string
  formatDate(
    value: Parameters<Intl.DateTimeFormat['format']>[0] | string,
    opts?: FormatDateOptions
  ): string
  formatTime(
    value: Parameters<Intl.DateTimeFormat['format']>[0] | string,
    opts?: FormatDateOptions
  ): string
  formatDateToParts(
    value: Parameters<Intl.DateTimeFormat['format']>[0] | string,
    opts?: FormatDateOptions
  ): Intl.DateTimeFormatPart[]
  formatTimeToParts(
    value: Parameters<Intl.DateTimeFormat['format']>[0] | string,
    opts?: FormatDateOptions
  ): Intl.DateTimeFormatPart[]
  formatRelativeTime(
    value: Parameters<Intl.RelativeTimeFormat['format']>[0],
    unit?: Parameters<Intl.RelativeTimeFormat['format']>[1],
    opts?: FormatRelativeTimeOptions
  ): string
  formatNumber(
    value: Parameters<Intl.NumberFormat['format']>[0],
    opts?: FormatNumberOptions
  ): string
  formatNumberToParts(
    value: Parameters<Intl.NumberFormat['format']>[0],
    opts?: FormatNumberOptions
  ): Intl.NumberFormatPart[]
  formatPlural(
    value: Parameters<Intl.PluralRules['select']>[0],
    opts?: FormatPluralOptions
  ): ReturnType<Intl.PluralRules['select']>
  formatMessage(
    descriptor: MessageDescriptor,
    values?: Record<string, PrimitiveType | FormatXMLElementFn<string, string>>,
    opts?: IntlMessageFormatOptions
  ): string
  formatMessage(
    descriptor: MessageDescriptor,
    values?: Record<string, PrimitiveType | T | FormatXMLElementFn<T, R>>,
    opts?: IntlMessageFormatOptions
  ): R
  $t(
    descriptor: MessageDescriptor,
    values?: Record<string, PrimitiveType | FormatXMLElementFn<string, string>>,
    opts?: IntlMessageFormatOptions
  ): string
  $t(
    descriptor: MessageDescriptor,
    values?: Record<string, PrimitiveType | T | FormatXMLElementFn<T, R>>,
    opts?: IntlMessageFormatOptions
  ): R
  formatList(values: ReadonlyArray<string>, opts?: FormatListOptions): string
  formatList(
    values: ReadonlyArray<string | T>,
    opts?: FormatListOptions
  ): T | string | Array<string | T>
  formatListToParts(
    values: ReadonlyArray<string | T>,
    opts?: FormatListOptions
  ): Part[]
  formatDisplayName(
    value: Parameters<DisplayNames['of']>[0],
    opts: FormatDisplayNameOptions
  ): string | undefined
}

export interface Formatters {
  getDateTimeFormat(
    ...args: ConstructorParameters<typeof Intl.DateTimeFormat>
  ): DateTimeFormat
  getNumberFormat(
    locales?: string | string[],
    opts?: NumberFormatOptions
  ): Intl.NumberFormat
  getMessageFormat(
    ...args: ConstructorParameters<typeof IntlMessageFormat>
  ): IntlMessageFormat
  getRelativeTimeFormat(
    ...args: ConstructorParameters<typeof Intl.RelativeTimeFormat>
  ): Intl.RelativeTimeFormat
  getPluralRules(
    ...args: ConstructorParameters<typeof Intl.PluralRules>
  ): Intl.PluralRules
  getListFormat(
    ...args: ConstructorParameters<typeof IntlListFormat>
  ): IntlListFormat
  getDisplayNames(
    ...args: ConstructorParameters<typeof DisplayNames>
  ): DisplayNames
}

export interface IntlShape<T = string>
  extends ResolvedIntlConfig<T>,
    IntlFormatters {
  formatters: Formatters
}

export interface IntlCache {
  dateTime: Record<string, DateTimeFormat>
  number: Record<string, Intl.NumberFormat>
  message: Record<string, IntlMessageFormat>
  relativeTime: Record<string, Intl.RelativeTimeFormat>
  pluralRules: Record<string, Intl.PluralRules>
  list: Record<string, IntlListFormat>
  displayNames: Record<string, DisplayNames>
}

export interface MessageDescriptor {
  id?: MessageIds
  description?: string | object
  defaultMessage?: string | MessageFormatElement[]
}

export type IntlConfig<T = string> = Omit<
  ResolvedIntlConfig<T>,
  keyof typeof DEFAULT_INTL_CONFIG
> &
  Partial<typeof DEFAULT_INTL_CONFIG>
