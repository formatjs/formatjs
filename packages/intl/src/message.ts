import {invariant} from '@formatjs/ecma402-abstract'

import {Formatters, MessageDescriptor, CustomFormats, OnErrorFn} from './types'

import {
  IntlMessageFormat,
  FormatXMLElementFn,
  PrimitiveType,
  Formatters as IntlMessageFormatFormatters,
  Options,
} from 'intl-messageformat'
import {MissingTranslationError, MessageFormatError} from './error'
import {TYPE, MessageFormatElement} from '@formatjs/icu-messageformat-parser'

function setTimeZoneInOptions(
  opts: Record<string, Intl.DateTimeFormatOptions>,
  timeZone: string
): Record<string, Intl.DateTimeFormatOptions> {
  return Object.keys(opts).reduce(
    (all: Record<string, Intl.DateTimeFormatOptions>, k) => {
      all[k] = {
        timeZone,
        ...opts[k],
      }
      return all
    },
    {}
  )
}

function deepMergeOptions(
  opts1: Record<string, Intl.DateTimeFormatOptions>,
  opts2: Record<string, Intl.DateTimeFormatOptions>
): Record<string, Intl.DateTimeFormatOptions> {
  const keys = Object.keys({...opts1, ...opts2})
  return keys.reduce((all: Record<string, Intl.DateTimeFormatOptions>, k) => {
    all[k] = {
      ...(opts1[k] || {}),
      ...(opts2[k] || {}),
    }
    return all
  }, {})
}

function deepMergeFormatsAndSetTimeZone(
  f1: CustomFormats,
  timeZone?: string
): CustomFormats {
  if (!timeZone) {
    return f1
  }
  const mfFormats = IntlMessageFormat.formats
  return {
    ...mfFormats,
    ...f1,
    date: deepMergeOptions(
      setTimeZoneInOptions(mfFormats.date, timeZone),
      setTimeZoneInOptions(f1.date || {}, timeZone)
    ),
    time: deepMergeOptions(
      setTimeZoneInOptions(mfFormats.time, timeZone),
      setTimeZoneInOptions(f1.time || {}, timeZone)
    ),
  }
}

export function formatMessage(
  config: {
    locale: string
    timeZone?: string
    formats: CustomFormats
    messages: Record<string, string> | Record<string, MessageFormatElement[]>
    defaultLocale: string
    defaultFormats: CustomFormats
    onError: OnErrorFn
  },
  state: Formatters,
  messageDescriptor?: MessageDescriptor,
  values?: Record<string, PrimitiveType | FormatXMLElementFn<string, string>>,
  opts?: Options
): string
export function formatMessage<T>(
  {
    locale,
    formats,
    messages,
    defaultLocale,
    defaultFormats,
    fallbackOnEmptyString,
    onError,
    timeZone,
    defaultRichTextElements,
  }: {
    locale: string
    timeZone?: string
    formats: CustomFormats
    messages: Record<string, string> | Record<string, MessageFormatElement[]>
    defaultLocale: string
    defaultFormats: CustomFormats
    defaultRichTextElements?: Record<string, FormatXMLElementFn<T>>
    fallbackOnEmptyString?: boolean
    onError: OnErrorFn
  },
  state: IntlMessageFormatFormatters & Pick<Formatters, 'getMessageFormat'>,
  messageDescriptor: MessageDescriptor = {id: ''},
  values?: Record<string, PrimitiveType | T | FormatXMLElementFn<T>>,
  opts?: Options
): Array<T | string> | string | T {
  const {id: msgId, defaultMessage} = messageDescriptor

  // `id` is a required field of a Message Descriptor.
  invariant(
    !!msgId,
    `[@formatjs/intl] An \`id\` must be provided to format a message. You can either:
1. Configure your build toolchain with [babel-plugin-formatjs](https://formatjs.io/docs/tooling/babel-plugin)
or [@formatjs/ts-transformer](https://formatjs.io/docs/tooling/ts-transformer) OR
2. Configure your \`eslint\` config to include [eslint-plugin-formatjs](https://formatjs.io/docs/tooling/linter#enforce-id)
to autofix this issue`
  )
  const id = String(msgId)
  const message =
    // In case messages is Object.create(null)
    // e.g import('foo.json') from webpack)
    // See https://github.com/formatjs/formatjs/issues/1914
    messages &&
    Object.prototype.hasOwnProperty.call(messages, id) &&
    messages[id]

  // IMPORTANT: Hot path if `message` is AST with a single literal node
  if (
    Array.isArray(message) &&
    message.length === 1 &&
    message[0].type === TYPE.literal
  ) {
    return message[0].value
  }

  // IMPORTANT: Hot path straight lookup for performance
  if (
    !values &&
    message &&
    typeof message === 'string' &&
    !defaultRichTextElements
  ) {
    return message.replace(/'\{(.*?)\}'/gi, `{$1}`)
  }
  values = {
    ...defaultRichTextElements,
    ...(values || {}),
  }
  formats = deepMergeFormatsAndSetTimeZone(formats, timeZone)
  defaultFormats = deepMergeFormatsAndSetTimeZone(defaultFormats, timeZone)

  if (!message) {
    if (fallbackOnEmptyString === false && message === '') {
      return message
    }

    if (
      !defaultMessage ||
      (locale && locale.toLowerCase() !== defaultLocale.toLowerCase())
    ) {
      // This prevents warnings from littering the console in development
      // when no `messages` are passed into the <IntlProvider> for the
      // default locale.
      onError(new MissingTranslationError(messageDescriptor, locale))
    }
    if (defaultMessage) {
      try {
        const formatter = state.getMessageFormat(
          defaultMessage,
          defaultLocale,
          defaultFormats,
          opts
        )

        return formatter.format(values)
      } catch (e) {
        onError(
          new MessageFormatError(
            `Error formatting default message for: "${id}", rendering default message verbatim`,
            locale,
            messageDescriptor,
            e
          )
        )
        return typeof defaultMessage === 'string' ? defaultMessage : id
      }
    }
    return id
  }

  // We have the translated message
  try {
    const formatter = state.getMessageFormat(message, locale, formats, {
      formatters: state,
      ...(opts || {}),
    })

    return formatter.format<T>(values)
  } catch (e) {
    onError(
      new MessageFormatError(
        `Error formatting message: "${id}", using ${
          defaultMessage ? 'default message' : 'id'
        } as fallback.`,
        locale,
        messageDescriptor,
        e
      )
    )
  }

  if (defaultMessage) {
    try {
      const formatter = state.getMessageFormat(
        defaultMessage,
        defaultLocale,
        defaultFormats,
        opts
      )

      return formatter.format(values)
    } catch (e) {
      onError(
        new MessageFormatError(
          `Error formatting the default message for: "${id}", rendering message verbatim`,
          locale,
          messageDescriptor,
          e
        )
      )
    }
  }

  if (typeof message === 'string') {
    return message
  }
  if (typeof defaultMessage === 'string') {
    return defaultMessage
  }
  return id
}
