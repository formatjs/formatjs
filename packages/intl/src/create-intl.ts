import {IntlCache, IntlShape, IntlConfig, ResolvedIntlConfig} from './types'
import {createFormatters, DEFAULT_INTL_CONFIG} from './utils'
import {InvalidConfigError, MissingDataError} from './error'
import {formatNumber, formatNumberToParts} from './number'
import {formatRelativeTime} from './relativeTime'
import {
  formatDate,
  formatDateToParts,
  formatTime,
  formatTimeToParts,
  formatDateTimeRange,
} from './dateTime'
import {formatPlural} from './plural'
import {formatMessage} from './message'
import {formatList, formatListToParts} from './list'
import {formatDisplayName} from './displayName'
import {MessageFormatElement} from '@formatjs/icu-messageformat-parser'

export interface CreateIntlFn<
  T = string,
  C extends IntlConfig<T> = IntlConfig<T>,
  S extends IntlShape<T> = IntlShape<T>
> {
  (config: C, cache?: IntlCache): S
}

function messagesContainString(
  messages: Record<string, string> | Record<string, MessageFormatElement[]>
): messages is Record<string, MessageFormatElement[]> {
  const firstMessage = messages ? messages[Object.keys(messages)[0]] : undefined

  return typeof firstMessage === 'string'
}

function verifyConfigMessages<T = string>(config: IntlConfig<T>) {
  if (
    config.onWarn &&
    config.defaultRichTextElements &&
    messagesContainString(config.messages || {})
  ) {
    config.onWarn(`[@formatjs/intl] "defaultRichTextElements" was specified but "message" was not pre-compiled. 
Please consider using "@formatjs/cli" to pre-compile your messages for performance.
For more details see https://formatjs.io/docs/getting-started/message-distribution`)
  }
}

/**
 * Create intl object
 * @param config intl config
 * @param cache cache for formatter instances to prevent memory leak
 */
export function createIntl<T = string>(
  config: IntlConfig<T>,
  cache?: IntlCache
): IntlShape<T> {
  const formatters = createFormatters(cache)
  const resolvedConfig: ResolvedIntlConfig<T> = {
    ...DEFAULT_INTL_CONFIG,
    ...config,
  }

  const {locale, defaultLocale, onError} = resolvedConfig
  if (!locale) {
    if (onError) {
      onError(
        new InvalidConfigError(
          `"locale" was not configured, using "${defaultLocale}" as fallback. See https://formatjs.io/docs/react-intl/api#intlshape for more details`
        )
      )
    }
    // Since there's no registered locale data for `locale`, this will
    // fallback to the `defaultLocale` to make sure things can render.
    // The `messages` are overridden to the `defaultProps` empty object
    // to maintain referential equality across re-renders. It's assumed
    // each <FormattedMessage> contains a `defaultMessage` prop.
    resolvedConfig.locale = resolvedConfig.defaultLocale || 'en'
  } else if (!Intl.NumberFormat.supportedLocalesOf(locale).length && onError) {
    onError(
      new MissingDataError(
        `Missing locale data for locale: "${locale}" in Intl.NumberFormat. Using default locale: "${defaultLocale}" as fallback. See https://formatjs.io/docs/react-intl#runtime-requirements for more details`
      )
    )
  } else if (
    !Intl.DateTimeFormat.supportedLocalesOf(locale).length &&
    onError
  ) {
    onError(
      new MissingDataError(
        `Missing locale data for locale: "${locale}" in Intl.DateTimeFormat. Using default locale: "${defaultLocale}" as fallback. See https://formatjs.io/docs/react-intl#runtime-requirements for more details`
      )
    )
  }

  verifyConfigMessages(resolvedConfig)
  return {
    ...resolvedConfig,
    formatters,
    formatNumber: formatNumber.bind(
      null,
      resolvedConfig,
      formatters.getNumberFormat
    ),
    formatNumberToParts: formatNumberToParts.bind(
      null,
      resolvedConfig,
      formatters.getNumberFormat
    ),
    formatRelativeTime: formatRelativeTime.bind(
      null,
      resolvedConfig,
      formatters.getRelativeTimeFormat
    ),
    formatDate: formatDate.bind(
      null,
      resolvedConfig,
      formatters.getDateTimeFormat
    ),
    formatDateToParts: formatDateToParts.bind(
      null,
      resolvedConfig,
      formatters.getDateTimeFormat
    ),
    formatTime: formatTime.bind(
      null,
      resolvedConfig,
      formatters.getDateTimeFormat
    ),
    formatDateTimeRange: formatDateTimeRange.bind(
      null,
      resolvedConfig,
      formatters.getDateTimeFormat
    ),
    formatTimeToParts: formatTimeToParts.bind(
      null,
      resolvedConfig,
      formatters.getDateTimeFormat
    ),
    formatPlural: formatPlural.bind(
      null,
      resolvedConfig,
      formatters.getPluralRules
    ),
    // @ts-expect-error TODO: will get to this later
    formatMessage: formatMessage.bind(null, resolvedConfig, formatters),
    // @ts-expect-error TODO: will get to this later
    $t: formatMessage.bind(null, resolvedConfig, formatters),
    formatList: formatList.bind(null, resolvedConfig, formatters.getListFormat),
    formatListToParts: formatListToParts.bind(
      null,
      resolvedConfig,
      formatters.getListFormat
    ),
    formatDisplayName: formatDisplayName.bind(
      null,
      resolvedConfig,
      formatters.getDisplayNames
    ),
  }
}
