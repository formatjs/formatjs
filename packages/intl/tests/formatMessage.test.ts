/* eslint-disable @typescript-eslint/camelcase */
import '@formatjs/intl-numberformat/polyfill'
import '@formatjs/intl-numberformat/locale-data/en'
import '@formatjs/intl-numberformat/locale-data/es'
import IntlMessageFormat from 'intl-messageformat'
import {parse} from '@formatjs/icu-messageformat-parser'
import {formatMessage as baseFormatMessage} from '../src/message'
import {Formatters, IntlConfig, IntlFormatters} from '../src/types'

describe('format API', () => {
  const {NODE_ENV} = process.env

  let config: IntlConfig<any>
  let state: Formatters

  beforeEach(() => {
    config = {
      locale: 'en',

      messages: {
        no_args: 'Hello, World!',
        with_arg: 'Hello, {name}!',
        with_named_format: 'It is {now, date, year_only}',
        with_html: 'Hello, <b>{name}</b>!',

        missing: undefined as any,
        empty: '',
        invalid: 'invalid {}',
        missing_value: 'missing {arg_missing}',
        missing_named_format: 'missing {now, date, format_missing}',
        richText: 'rich <b>text</b>',
        ast_simple: parse('hello world'),
        ast_var: parse('hello there, {name}'),
      } as Record<string, any>,

      formats: {
        date: {
          'year-only': {
            year: 'numeric',
          },
          missing: undefined,
        },

        time: {
          'hour-only': {
            hour: '2-digit',
            hour12: false,
          },
          missing: undefined,
        },

        relative: {
          seconds: {
            style: 'narrow',
          },
          missing: undefined,
        },

        number: {
          percent: {
            style: 'percent',
            minimumFractionDigits: 2,
          },
          missing: undefined,
        },
      } as any,

      defaultLocale: 'en',
      defaultFormats: {},

      onError: jest.fn(),
    }

    state = {
      getDateTimeFormat: jest
        .fn()
        .mockImplementation((...args) => new Intl.DateTimeFormat(...args)),
      getNumberFormat: jest
        .fn()
        .mockImplementation((...args) => new Intl.NumberFormat(...args)),
      getMessageFormat: jest
        .fn()
        .mockImplementation(
          (msg, ...args) => new IntlMessageFormat(msg, ...args)
        ),
      getRelativeTimeFormat: jest
        .fn()
        .mockImplementation((...args) => new Intl.RelativeTimeFormat(...args)),
      getPluralRules: jest
        .fn()
        .mockImplementation((...args) => new Intl.PluralRules(...args)),
      getListFormat: jest
        .fn()
        .mockImplementation((...args) => new Intl.ListFormat(...args)),
      getDisplayNames: jest
        .fn()
        .mockImplementation(
          (...args) => new (Intl as any).DisplayNames(...args)
        ),
    }
  })

  afterEach(() => {
    process.env.NODE_ENV = NODE_ENV
  })

  describe('formatMessage()', () => {
    let formatMessage: IntlFormatters['formatMessage']

    beforeEach(() => {
      // @ts-ignore
      formatMessage = baseFormatMessage.bind(null, config, state)
    })
    it('should hot path message without values', function () {
      ;(state.getMessageFormat as jest.Mock).mockClear()
      expect(formatMessage({id: 'no_args'})).toBe('Hello, World!')
      expect(state.getMessageFormat).not.toHaveBeenCalled()
      expect(formatMessage({id: 'with_arg'}, {name: 'foo'})).toBe('Hello, foo!')
      expect(state.getMessageFormat).toHaveBeenCalled()
    })
    it('should hot path message without values', function () {
      ;(state.getMessageFormat as jest.Mock).mockClear()
      const err = jest.spyOn(console, 'error')
      expect(formatMessage({id: 'no_args'})).toBe('Hello, World!')
      expect(err).not.toHaveBeenCalled()
    })
    it('should not crash of messages does not have Object.prototype', function () {
      const messages = Object.create(null)
      messages!.no_args = 'Hello'
      // @ts-ignore
      formatMessage = baseFormatMessage.bind(
        null,
        {
          ...config,
          messages,
        },
        state
      )
      expect(() => formatMessage({id: 'no_args'})).not.toThrow()
      expect(formatMessage({id: 'no_args'})).toBe('Hello')
    })
    ;[`Hello, World!'{foo}'`, `'\ud83d'\udc04`].forEach(msg =>
      it(`should render escaped msg ${msg} properly in production`, () => {
        process.env.NODE_ENV = 'production'

        const descriptor = {
          id: 'hello',
          defaultMessage: msg,
        }

        const mf = new IntlMessageFormat(msg, 'en')

        expect(formatMessage(descriptor)).toBe(mf.format())
      })
    )

    it('throws when no Message Descriptor is provided', () => {
      // @ts-ignore
      expect(() => formatMessage()).toThrow(
        '[@formatjs/intl] An `id` must be provided to format a message.'
      )
    })

    it('throws when Message Descriptor `id` is missing or falsy', () => {
      expect(() => formatMessage({})).toThrow(
        '[@formatjs/intl] An `id` must be provided to format a message.'
      )
      ;[undefined, null, false, 0, ''].forEach(id => {
        // @ts-ignore
        expect(() => formatMessage({id})).toThrow(
          '[@formatjs/intl] An `id` must be provided to format a message.'
        )
      })
    })

    it('formats basic messages', () => {
      const {locale, messages} = config
      const mf = new IntlMessageFormat(messages!.no_args, locale)

      expect(formatMessage({id: 'no_args'})).toBe(mf.format())
    })

    it('formats basic message with preparsed defaultMessage', () => {
      const {locale, messages} = config
      const mf = new IntlMessageFormat(messages!.ast_var, locale)

      expect(
        formatMessage(
          {id: 'foo', defaultMessage: messages!.ast_var},
          {
            name: 'hey',
          }
        )
      ).toBe(
        mf.format({
          name: 'hey',
        })
      )
    })

    it('formats message with ID as a method in Object.prototype, GH issue #1885', () => {
      expect(formatMessage({id: 'toString'})).toBe('toString')
      expect(formatMessage({id: '__proto__'})).toBe('__proto__')
    })

    it('formats legacy HTML messages', () => {
      const {locale, messages} = config
      const mf = new IntlMessageFormat(messages!.richText, locale)
      const values = {
        b: (s: string) => `<foobar>${s}</foobar>`,
      }
      expect(formatMessage({id: 'richText'}, values)).toBe(
        // @ts-ignore
        mf.format<string>(values)
      )
    })

    it('formats basic AST messages', () => {
      const {locale, messages} = config
      const mf = new IntlMessageFormat(messages!.ast_simple, locale)

      expect(formatMessage({id: 'ast_simple'})).toBe(mf.format())
    })

    it('formats basic AST messages in prod', () => {
      const {locale, messages} = config
      const mf = new IntlMessageFormat(messages!.ast_simple, locale)
      process.env.NODE_ENV = 'production'
      expect(formatMessage({id: 'ast_simple'})).toBe(mf.format())
    })

    it('formats messages with placeholders', () => {
      const {locale, messages} = config
      const mf = new IntlMessageFormat(messages!.with_arg, locale)
      const values = {name: 'Eric'}

      expect(formatMessage({id: 'with_arg'}, values)).toBe(mf.format(values))
    })

    it('formats AST message with placeholders', () => {
      const {locale, messages} = config
      const mf = new IntlMessageFormat(messages!.ast_var, locale)
      const values = {name: 'Eric'}

      expect(formatMessage({id: 'ast_var'}, values)).toBe(mf.format(values))
    })

    it('formats messages with named formats', () => {
      const {locale, messages, formats} = config
      const mf = new IntlMessageFormat(
        messages!.with_named_format,
        locale,
        formats
      )
      const values = {now: Date.now()}

      expect(formatMessage({id: 'with_named_format'}, values)).toBe(
        mf.format(values)
      )
    })

    describe('fallbacks', () => {
      it('formats message with missing named formats', () => {
        const {locale, messages} = config
        const mf = new IntlMessageFormat(messages!.missing_named_format, locale)
        const values = {now: Date.now()}

        expect(formatMessage({id: 'missing_named_format'}, values)).toBe(
          mf.format(values)
        )
      })

      it('formats `defaultMessage` when message is missing', () => {
        const {locale, messages} = config
        const mf = new IntlMessageFormat(messages!.with_arg, locale)
        const id = 'missing'
        const values = {name: 'Eric'}

        expect(
          formatMessage(
            {
              id: id,
              defaultMessage: messages!.with_arg,
            },
            values
          )
        ).toBe(mf.format(values))
      })

      it('warns when `message` is missing and locales are different', () => {
        config.locale = 'fr'

        const {locale, messages, defaultLocale} = config
        const mf = new IntlMessageFormat(messages!.with_arg, locale)
        const id = 'missing'
        const values = {name: 'Eric'}

        expect(locale).not.toEqual(defaultLocale)

        expect(
          formatMessage(
            {
              id,
              defaultMessage: messages!.with_arg,
            },
            values
          )
        ).toBe(mf.format(values))

        expect((config.onError as jest.Mock).mock.calls.map(c => c[0].code))
          .toMatchInlineSnapshot(`
          Array [
            "MISSING_TRANSLATION",
          ]
        `)
      })

      it('warns when `message` and `defaultMessage` are missing', () => {
        const {messages} = config
        const id = 'missing'
        const values = {name: 'Eric'}

        expect(
          formatMessage(
            {
              id: id,
              defaultMessage: messages!.missing,
            },
            values
          )
        ).toBe(id)

        expect((config.onError as jest.Mock).mock.calls.map(c => c[0].code))
          .toMatchInlineSnapshot(`
          Array [
            "MISSING_TRANSLATION",
          ]
        `)
      })

      it('formats `defaultMessage` when message has a syntax error', () => {
        const {locale, messages} = config
        const mf = new IntlMessageFormat(messages!.with_arg, locale)
        const id = 'invalid'
        const values = {name: 'Eric'}

        expect(
          formatMessage(
            {
              id: id,
              defaultMessage: messages!.with_arg,
            },
            values
          )
        ).toBe(mf.format(values))

        expect(
          (config.onError as jest.Mock).mock.calls.map(c => c[0].code)
        ).toMatchSnapshot()
      })

      it('formats `defaultMessage` when message has missing values', () => {
        const {locale, messages} = config
        const mf = new IntlMessageFormat(messages!.with_arg, locale)
        const id = 'missing_value'
        const values = {name: 'Eric'}

        expect(
          formatMessage(
            {
              id: id,
              defaultMessage: messages!.with_arg,
            },
            values
          )
        ).toBe(mf.format(values))

        expect(
          (config.onError as jest.Mock).mock.calls.map(c => c[0].code)
        ).toMatchSnapshot()
      })

      it('returns message source when message and `defaultMessage` have formatting errors', () => {
        const {messages} = config
        const id = 'missing_value'

        expect(
          formatMessage(
            {
              id,
              defaultMessage: messages!.invalid,
            },
            {
              foo: 1,
            }
          )
        ).toBe(messages![id])
        expect(
          (config.onError as jest.Mock).mock.calls.map(c => c[0].code)
        ).toMatchSnapshot()
      })

      it('returns message source when formatting error and missing `defaultMessage`', () => {
        const {messages} = config
        const id = 'missing_value'

        expect(
          formatMessage(
            {
              id,
              defaultMessage: messages!.missing,
            },
            {foo: 1}
          )
        ).toBe(messages![id])
        expect(
          (config.onError as jest.Mock).mock.calls.map(c => c[0].code)
        ).toMatchSnapshot()
      })

      it('returns `defaultMessage` source when formatting errors and missing message', () => {
        config.locale = 'en-US'

        const {messages} = config
        const id = 'missing'

        expect(
          formatMessage({
            id,
            defaultMessage: messages!.invalid,
          })
        ).toBe(messages!.invalid)

        expect(
          (config.onError as jest.Mock).mock.calls.map(c => c[0].code)
        ).toMatchSnapshot()
      })

      it('returns message `id` when message and `defaultMessage` are missing', () => {
        const id = 'missing'

        expect(formatMessage({id})).toBe(id)

        expect(
          (config.onError as jest.Mock).mock.calls.map(c => c[0].code)
        ).toMatchSnapshot()
      })

      it('returns an empty string when `fallbackOnEmptyString` is false', () => {
        config.fallbackOnEmptyString = false
        const id = 'empty'

        expect(
          formatMessage({ id })
        ).toBe('')
      })

      it('does not return an empty string when `fallbackOnEmptyString` is true', () => {
        config.fallbackOnEmptyString = true
        const id = 'empty'

        expect(
          formatMessage({ id })
        ).toBe(id)
      })

      it('returns message `id` when message and `defaultMessage` are empty', () => {
        const {messages} = config
        const id = 'empty'

        expect(
          formatMessage({
            id: id,
            defaultMessage: messages![id],
          })
        ).toBe(id)

        expect(
          (config.onError as jest.Mock).mock.calls.map(c => c[0].code)
        ).toMatchSnapshot()
      })

      it('allow passing Intl.MessageFormat opts in', function () {
        const {locale, messages, formats} = config
        const opts = {
          ignoreTag: true,
        }
        const mf = new IntlMessageFormat(
          messages!.richText,
          locale,
          formats,
          opts
        )

        expect(formatMessage({id: 'richText'}, opts)).toBe(mf.format())
      })
    })
  })
})
