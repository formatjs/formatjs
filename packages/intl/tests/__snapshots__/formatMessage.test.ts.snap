// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format API formatMessage() fallbacks formats \`defaultMessage\` when message has a syntax error 1`] = `
[
  "FORMAT_ERROR",
]
`;

exports[`format API formatMessage() fallbacks formats \`defaultMessage\` when message has missing values 1`] = `
[
  "FORMAT_ERROR",
]
`;

exports[`format API formatMessage() fallbacks returns \`defaultMessage\` source when formatting errors and missing message 1`] = `
[
  "MISSING_TRANSLATION",
  "FORMAT_ERROR",
]
`;

exports[`format API formatMessage() fallbacks returns message \`id\` when message and \`defaultMessage\` are empty 1`] = `
[
  "MISSING_TRANSLATION",
]
`;

exports[`format API formatMessage() fallbacks returns message \`id\` when message and \`defaultMessage\` are missing 1`] = `
[
  "MISSING_TRANSLATION",
]
`;

exports[`format API formatMessage() fallbacks returns message source when formatting error and missing \`defaultMessage\` 1`] = `
[
  "FORMAT_ERROR",
]
`;

exports[`format API formatMessage() fallbacks returns message source when message and \`defaultMessage\` have formatting errors 1`] = `
[
  "FORMAT_ERROR",
  "FORMAT_ERROR",
]
`;
